src/main/java/vct/col/veymont/GenerateLTS.scala

```diff
@@ -196,44 +148,43 @@ class GenerateLTS(override val source : ProgramUnit, isGlobal : Boolean) extends
     nextState
   }
 
-  def weakSequenceAllowed(s1 : ASTNode, s2: ASTNode) : Boolean = {
-    if(!canSwap(s1) || !canSwap(s2))
-      false
-    else getSubjects(Set.empty,s1).intersect(getSubjects(Set.empty,s2)).isEmpty
-  }
+  def weakSequenceAllowed(s1 : ASTNode, s2: ASTNode) : Boolean =
+    canSwap(s1) && canSwap(s2) && getSubjects(Set.empty,s1).intersect(getSubjects(Set.empty,s2)).isEmpty
+
 
   def getSubjects(seen : Set[String],s : ASTNode) : Set[String] = {
     s match {
       case b : BlockStatement => b.getStatements.toSet.flatMap(getSubjects(seen,_))
-      case a : AssignmentStatement => getGlobalAction(a) match {
-        case SingleRoleAction(node) => node match {
-          case an : AssignmentStatement => an.location match {
-            case d : Dereference => Set(d.obj.asInstanceOf[NameExpression].name)
-            case n : NameExpression => Set(n.name)
-            case _ => throw Failure("VeyMont Fail: cannot determine subject of  assignment " + a.toString)
-          }
-          case m : MethodInvokation => m.`object` match {
-            case n : NameExpression => Set(n.name)
-            case _ => throw Failure("VeyMont Fail: cannot determine subject of  assignment " + a.toString)
-          }
-        }
-        case CommunicationAction(receiver,_, sender, _) =>
-          Set(receiver.name, sender.name)
-        case _ => Set.empty
-      }
+      case a : AssignmentStatement => getSubjectAssignment(a)
       case _ : IfStatement => roleNames.toSet
       case _ : LoopStatement => roleNames.toSet
       case p : ParallelRegion => p.blocks.map(_.block).toSet.flatMap(getSubjects(seen,_))
       case m : MethodInvokation => {
         if(m.`object` == null) { //it is a main method
-          Fail("VeyMont Fail: encountered method call %s in LTS generation",m.method)
-          Set.empty
```

Comment:
I see we use both Fail and throw Failure. What is our policy for throwing errors. When do we use which one?

Reply:
Is there a reason you changed it?

Reply:
`throw Failure` indicates to scala that the code does not terminate normally (i.e. always results in an exception), so you can use it in place of any value (note that this method returns something). Ideally we want the logging stuff to be written in scala, so we can indicate that `Fail` (the equivalent method) behaves in this way. That's not currently possible because it's written in Java. imo we want to use `Fail` whenever possible, but this usage is better for now in places where a value is expected, and hopefully we can soon use `Fail` in those places also.